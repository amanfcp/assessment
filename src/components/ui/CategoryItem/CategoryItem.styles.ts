import {StyleSheet} from 'react-native';

const styles = StyleSheet.create({
  item: {
    flexDirection: 'row',
    alignItems: 'center',
    marginVertical: 5,
    marginLeft: 12,
  },
  description: {
    fontSize: 10,
    marginLeft: 4,
  },
  subcategories: {
    marginLeft: 12,
  },
});

export default styles;
