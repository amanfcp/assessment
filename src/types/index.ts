export type Category = {
  id: number;
  name: string;
  description?: string;
  headId: number | null;
};

export type HierarchicalCategory = Category & {
  subCategories: HierarchicalCategory[];
};
