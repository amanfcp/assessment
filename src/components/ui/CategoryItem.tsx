import React from 'react';
import {StyleSheet, Text, View} from 'react-native';
import {HierarchicalCategory} from '../../types';
import CheckBox from '@react-native-community/checkbox';

type CategoryItemProps = {
  item: HierarchicalCategory;
  toggleCategory: (id: number) => void;
  selectedCategoryIds: number[];
};

export default function CategoryItem({
  item,
  toggleCategory,
  selectedCategoryIds,
}: CategoryItemProps) {
  const renderCategoryItem = (currentItem: HierarchicalCategory) => {
    const Item = (
      <View key={`category-${currentItem.id}`} style={styles.item}>
        <CheckBox
          value={selectedCategoryIds.some(cat => cat === currentItem.id)}
          onValueChange={() => toggleCategory(currentItem.id)}
        />
        <View>
          <Text>{currentItem.name}</Text>
          {currentItem.description && (
            <Text style={styles.description}>{currentItem.description}</Text>
          )}
        </View>
      </View>
    );

    if (currentItem.subCategories.length === 0) {
      return Item;
    }

    return (
      <View key={`category-${currentItem.id}`}>
        {Item}
        <View style={styles.subcategories}>
          {currentItem.subCategories.map(subcategory =>
            renderCategoryItem(subcategory),
          )}
        </View>
      </View>
    );
  };

  return renderCategoryItem(item);
}

const styles = StyleSheet.create({
  item: {
    flexDirection: 'row',
    alignItems: 'center',
    marginVertical: 5,
    marginLeft: 12,
  },
  description: {
    fontSize: 10,
    marginLeft: 4,
  },
  subcategories: {
    marginLeft: 12,
  },
});
