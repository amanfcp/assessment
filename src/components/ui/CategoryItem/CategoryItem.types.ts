import {HierarchicalCategory} from '../../../types';

export type CategoryItemProps = {
  item: HierarchicalCategory;
  toggleCategory: (id: number) => void;
  selectedCategoryIds: number[];
};
