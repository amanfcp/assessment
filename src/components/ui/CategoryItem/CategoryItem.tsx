import React from 'react';
import {Text, View} from 'react-native';
import {HierarchicalCategory} from '../../../types';
import CheckBox from '@react-native-community/checkbox';
import styles from './CategoryItem.styles';
import {CategoryItemProps} from './CategoryItem.types';

export default function CategoryItem({
  item,
  toggleCategory,
  selectedCategoryIds,
}: CategoryItemProps) {
  const renderCategoryItem = (currentItem: HierarchicalCategory) => {
    const Item = (
      <View key={`category-${currentItem.id}`} style={styles.item}>
        <CheckBox
          accessibilityLabel={`category-${currentItem.id}-checkbox`}
          value={selectedCategoryIds.some(cat => cat === currentItem.id)}
          onValueChange={() => toggleCategory(currentItem.id)}
        />
        <View>
          <Text>{currentItem.name}</Text>
          {currentItem.description && (
            <Text style={styles.description}>{currentItem.description}</Text>
          )}
        </View>
      </View>
    );

    if (!currentItem.subCategories.length) {
      return Item;
    }

    return (
      <View key={`category-${currentItem.id}`}>
        {Item}
        <View style={styles.subcategories}>
          {currentItem.subCategories.map(subcategory =>
            renderCategoryItem(subcategory),
          )}
        </View>
      </View>
    );
  };

  return renderCategoryItem(item);
}
