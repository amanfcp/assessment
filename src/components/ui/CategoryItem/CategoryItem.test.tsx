import * as React from 'react';
import CategoryItem from './CategoryItem';
import {HierarchicalCategory} from '../../../types';
import {fireEvent, render, screen} from '@testing-library/react-native';

describe('CategoryItem', () => {
  const item: HierarchicalCategory = {
    id: 1,
    name: 'Phones',
    headId: null,
    description: '',
    subCategories: [],
  };

  it(`renders correctly`, () => {
    const tree = render(
      <CategoryItem
        item={item}
        selectedCategoryIds={[]}
        toggleCategory={jest.fn()}
      />,
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders the category name correctly', () => {
    render(
      <CategoryItem
        item={item}
        selectedCategoryIds={[]}
        toggleCategory={jest.fn()}
      />,
    );

    expect(screen.getByText('Phones')).toBeTruthy();
  });

  it('calls toggleCategory when clicked', () => {
    const toggleCategory = jest.fn();

    render(
      <CategoryItem
        item={item}
        selectedCategoryIds={[]}
        toggleCategory={toggleCategory}
      />,
    );

    fireEvent(
      screen.getByLabelText(`category-${item.id}-checkbox`),
      'onValueChange',
      {nativeEvent: {}},
    );
    expect(toggleCategory).toHaveBeenCalled();
  });

  it('is selected when its ID is in selectedCategoryIds', () => {
    render(
      <CategoryItem
        item={item}
        selectedCategoryIds={[1]}
        toggleCategory={jest.fn()}
      />,
    );

    const checkbox = screen.getByLabelText(`category-${item.id}-checkbox`);
    expect(checkbox.props.value).toBeTruthy();
  });

  it('should render a category with a subcategory', () => {
    item.subCategories = [
      {
        id: 2,
        name: 'iPhone',
        headId: 1,
        description: '',
        subCategories: [],
      },
    ];

    render(
      <CategoryItem
        item={item}
        selectedCategoryIds={[]}
        toggleCategory={jest.fn()}
      />,
    );
    const subCategory = screen.getByText('iPhone');
    expect(subCategory).toBeTruthy();
  });
});
