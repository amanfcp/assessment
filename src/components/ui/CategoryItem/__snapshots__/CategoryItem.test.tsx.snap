// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CategoryItem renders correctly 1`] = `
<View
  style={
    {
      "alignItems": "center",
      "flexDirection": "row",
      "marginLeft": 12,
      "marginVertical": 5,
    }
  }
>
  <View
    accessibilityRole="checkbox"
    accessibilityState={
      {
        "checked": false,
        "disabled": false,
      }
    }
    accessibilityValue={
      {
        "text": "off",
      }
    }
    accessible={true}
    pointerEvents="auto"
  >
    <RNCCheckbox
      accessibilityLabel="category-1-checkbox"
      forwardedRef={null}
      onValueChange={[Function]}
      style={
        [
          {
            "backgroundColor": "transparent",
            "height": 32,
            "width": 32,
          },
          undefined,
        ]
      }
      value={false}
    />
  </View>
  <View>
    <Text>
      Phones
    </Text>
  </View>
</View>
`;
